src/connector/codec/src/decoder/json/mod.rs

```diff
@@ -12,11 +12,180 @@
+    }
+
+    async fn deref(
+        &mut self,
+        value: &mut Value,
+        base_url: &Url,
+        used_refs: &Vec<String>,
+    ) -> Result<()> {
+        if let Some(obj) = value.as_object_mut()
+            && let Some(ref_value) = obj.remove("$ref")
+            && let Some(ref_string) = ref_value.as_str()
+        {
+            let ref_url = base_url.join(ref_string).into_url_parse(ref_string)?;
+            let mut ref_url_no_fragment = ref_url.clone();
+            ref_url_no_fragment.set_fragment(None);
+            let url_schema = ref_url_no_fragment.scheme();
+            let ref_no_fragment = ref_url_no_fragment.to_string();
+
+            let mut schema = match self.schema_cache.get(&ref_no_fragment) {
+                Some(cached_schema) => cached_schema.clone(),
+                None => {
+                    if url_schema == "http" || url_schema == "https" {
+                        reqwest::get(ref_url_no_fragment.clone())
+                            .await
+                            .into_request(&ref_no_fragment)?
+                            .json()
+                            .await
+                            .into_request(&ref_no_fragment)?
+                    } else if url_schema == "file" {
+                        let file_path = ref_url_no_fragment.to_file_path().map_err(|_| {
+                            anyhow::anyhow!(
+                                "could not convert url {} to file path",
+                                &ref_url_no_fragment
+                            )
+                        })?;
+                        let file =
+                            fs::File::open(file_path).into_schema_from_file(&ref_no_fragment)?;
+                        serde_json::from_reader(file)
+                            .into_schema_not_json_serde(ref_no_fragment.clone())?
+                    } else {
+                        return Err(Error::UnsupportedUrl {
+                            url: ref_no_fragment,
+                        });
+                    }
+                }
+            };
+
+            if !self.schema_cache.contains_key(&ref_no_fragment) {
+                self.schema_cache
+                    .insert(ref_no_fragment.clone(), schema.clone());
+            }
+
+            let ref_url_string = ref_url.to_string();
+            if let Some(ref_fragment) = ref_url.fragment() {
+                schema = schema
+                    .pointer(ref_fragment)
+                    .ok_or(Error::JsonRefPointerNotFound {
+                        ref_string: ref_string.to_owned(),
+                        ref_fragment: ref_fragment.to_owned(),
+                    })?
+                    .clone();
+            }
+            // Do not deref a url twice to prevent infinite loops
+            if used_refs.contains(&ref_url_string) {
+                return Ok(());
+            }
+            let mut new_used_refs = used_refs.clone();
+            new_used_refs.push(ref_url_string);
```

Comment:
Shall we add a test case on recursion? This `new_used_refs` is now unused.

Reply:
You are right. I mis deleted 1 extra deref line. Test added.